src/main/kotlin/com/haroldadmin/cnradapter/NetworkResponse.kt

```diff
@@ -1,61 +1,105 @@
+     * If you expect your server response to not contain a body, set the success body type ([S]) to [Unit],
+     * or ensure that the response code is 204.
      *
-     *    is NetworkResponse.Error -> // Action failed do something with error
-     * }
+     * @param body The parsed body of the successful response.
+     * @param response The un-parsed [Response] from Retrofit
      */
-    interface Error {
-        val error: Throwable
-    }
+    public class Success<S, E>(
+        public val body: S,
+        public val response: Response<*>
+    ) : NetworkResponse<S, E> {
+        /**
+         * The status code returned by the server.
+         *
+         * Alias for [Response.code] of the original response
+         */
+        public val code: Int
+            get() = response.code()
 
-    /**
-     * A request that resulted in a response with a non-2xx status code.
-     */
-    data class ServerError<U : Any>(
-        val body: U?,
-        val code: Int,
-        val headers: Headers? = null
-    ) : NetworkResponse<Nothing, U>(), Error {
-        override val error = IOException("Network server error: $code \n$body")
+        /**
+         * The headers returned by the server.
+         *
+         * Alias for [Response.headers] of the original response
+         */
+        public val headers: Headers
+            get() = response.headers()
     }
 
     /**
-     * A request that didn't result in a response.
+     * The result of a failed network request.
      */
-    data class NetworkError(override val error: IOException) :
-        NetworkResponse<Nothing, Nothing>(), Error
+    public sealed interface Error<S, E> : NetworkResponse<S, E> {
+        public val body: E?
+        public val error: Throwable?
 
-    /**
-     * A request that resulted in an error different from an IO or Server error.
-     *
-     * An example of such an error is JSON parsing exception thrown by a serialization library.
-     */
-    data class UnknownError(
-        override val error: Throwable,
-        val code: Int? = null,
-        val headers: Headers? = null,
-    ) : NetworkResponse<Nothing, Nothing>(), Error
-}
\ No newline at end of file
+        /**
+         * The result of a non 2xx response to a network request
+         */
+        public data class ServerError<S, E>(
+            public override val body: E?,
+            public val response: Response<*>?,
+        ) : Error<S, E> {
+            /**
+             * The status code returned by the server.
+             *
+             * Alias for [Response.code] of the original response
+             */
+            public val code: Int? = response?.code()
+
+            /**
+             * The headers returned by the server.
+             *
+             * Alias for [Response.headers] of the original response
+             */
+            public val headers: Headers? = response?.headers()
+
+            override val error: Throwable? = null
```

Comment:
Missing comment, if you already commented everything, you might as well add a simple comment here as well